SPV multi-server cross-validation

shesek opened the pull request:
Also includes a bunch of other small changes/fixups.

This PR depends on:
- ~~A [fix](https://github.com/Blockstream/electrs/commit/ff9237df4223468d4f1260cc377a013cf3a1ed25) to Blockstream/electrs that isn't yet deployed to the live servers. The GitLab CI was updated to use the updated version of electrs. (https://github.com/Blockstream/gdk/pull/59/commits/8ddc1c8e53be12df988f21ea5eeb51ec8489ade9)~~
  This was deployed.

- ~~A rust-bitcoin fork with serializable uint types. This is needed to store the validation result in the cache. A PR was sent upstream: https://github.com/rust-bitcoin/rust-bitcoin/pull/511~~ Merged.

- ~~A rust-electrum-client fork with support for using custom `TcpStream`s. This is needed to set timeouts. A PR was sent upstream: https://github.com/bitcoindevkit/rust-electrum-client/pull/25~~
  Upgraded to upstream rust-electrum-client v0.4.0 with the necessary changes.

RCasatta commented on subprojects/gdk_rust/gdk_electrum/src/spv.rs:
```diff
@@ -0,0 +1,338 @@
+    let fork = get_fork_branch(chain, &client, remote_tip_height, None)?;
+
+    let our_work: Uint256 = (fork.common_ancestor + 1..=chain.height())
+        .fold(Uint256::zero(), |total, height| total + chain.get(height).unwrap().work());
```
I think it's better to remove this unwrap, either by removing the closure or using try_fold?

shesek replied:
It is a logical error for any of these blocks not to exists, since it iterates up to a upper bound reported by HeadersChain as its tip height. If they are somehow missing, it would indicate that the state got irrevocably corrupted, in which case I think that panicking and crashing the whole thing down is a reasonable thing to do.

RCasatta replied:
why not unwrapping at line 158 then?
I have no strong position but I am not convinced either, some file system error for example could be treated by catching the error and deleting the file

shesek replied:
> why not unwrapping at line 158 then?

I would actually change line 158 to `unwrap()` too.

Another reason for this is that I defined `CrossValidationError` to mean an error that indicates something is wrong with the secondary server and not with the primary server/chain, so its kinda strange that a missing expected local header would result in a variant of `CrossValidationError`. Though the definition could of course be changed if deemed necessary.

> some file system error for example could be treated by catching the error and deleting the file

Wouldn't you also need to restore the file's contents after deleting it?

But in any case, until we have the error recovery logic implemented, if we somehow end up with corrupted database state and the choice is between letting the program continue doing "business as usual" without a specialized recovery procedure or crashing it, I would opt to crash it.

RCasatta replied:
ok, fine

shesek commented:
I added a new `SPVVerifyResult::NotLongest` status. Its encoded as the string `not_longest` or as the number `4`. When spv cross-validation fails, all of the transactions will report this status.

Should this status also be exposed as a global flag of some sort, or is having this on the individual transactions sufficient?

The cross-validator reports some information about the chain fork, including how far behind the primary electrum server is compared to the longest chain (both work and height wise). Should this information be made available to the app?

shesek commented:
I added a script for updating the Electrum servers list from the latest Electrum release: https://github.com/shesek/gdk/blob/202010-spv-crossvalidation/subprojects/gdk_rust/gdk_electrum/scripts/update-server-list.sh

Should this be added to some build process somewhere?

shesek commented:
Rebased on master and removed the dependency on my `rust-electrum-client` fork (the necessary changes were added upstream and released in v0.4.0). It still depends on my `rust-bitcoin` fork pending a PR I sent there.

Other standing items/questions:

- The apps need to handle the new `NotLongest` and `Unconfirmed` SPV verification statuses, as well as support configuring the new `spv_cross_validation` and `spv_cross_validation_servers` options.

- Should the SPV cross-validation status be exposed as global session state, in addition to being included on individual transactions? 

  This could be useful for displaying cross-validation failures somewhere prominent (outside of individual txs), and could also provide the app some additional details about the chain fork, like the local height, the height of the best known chain and the additional work in the best chain.

- Cross-validation is still not implemented in the C FFI used by the Green apps for SPV verification. In that environment, the Rust side doesn't run background threads of its own, it only runs when invoked from the CPP side. This introduces some complexities, both because cross-validation is an ongoing process and because the secondary servers are likely to be slow (or entirely unresponsive), which could result in user-visible latency if its not done in a separate thread.

LeoComandini commented on src/network_parameters.cpp:
```diff
@@ -78,8 +78,8 @@ static std::map<std::string, std::shared_ptr<nlohmann::json>> registered_network
-                { "electrum_url", "localhost:19002" }, { "spv_enabled", false }, { "tls", false },
-                { "server_type", "green" } })) },
+                { "electrum_url", "localhost:19002" }, { "tls", false }, { "server_type", "green" },
+                { "spv_enabled", false }, { "spv_cross_validation", false }, { "spv_cross_validation_servers", nlohmann::json::array() } })) },
```
Please run `./tools/format.sh` to format `src/network_parameters.cpp` correctly.

I recently added https://github.com/Blockstream/gdk/commit/ac593ff6c31948eef99b95068100ae81a81b5499 so that they are run in travis as well

shesek replied:
I wasn't aware this existed, will do.

shesek replied:
Formatted in a914e561eae54abb6b7b5e7d853a66cc32df1d48.

LeoComandini commented on subprojects/gdk_rust/Cargo.toml:
```diff
@@ -39,3 +39,7 @@ panic = 'abort'
+
+[patch.crates-io]
+bitcoin = { git = "https://github.com/shesek/rust-bitcoin", branch = "0.25.2-uint-serde", features = [ "use-serde" ] }
+electrum-client = { git = "https://github.com/shesek/rust-electrum-client", branch = "202011-custom-ssl-stream" }
```
Do this need to be different from `electrum-client = "0.4.0-beta.1"`?

shesek replied:
It used to be needed but the necessary changes were merged upstream, I updated `gdk_rust/gdk_electrum/Cargo.toml` in https://github.com/Blockstream/gdk/pull/59/commits/a71aae7ec1ea645bd7361a9606a4ee082e0a7b0d but forgot about `gdk_rust/Cargo.toml`. Will update this too.

shesek replied:
Updated (forced-pushed over the previous commit)

LeoComandini commented on subprojects/gdk_rust/gdk_electrum/src/servers-mainnet.txt:
```diff
@@ -0,0 +1,62 @@
+electrum.aantonop.com:50002:s:noverify
+electrum.bitkoins.nl:50512:s:noverify
+electrum.blockstream.info:50002:s:noverify
+blockstream.info:700:s:noverify
```
These are not ordered, so I guess they were obtained with a old version of the script, but I cannot verify them.

Also they change from different Electrum versions, we might want to set such version somewhere for verifiers

shesek replied:
Yes, I added sorting at a later stage without updating the server list.

I updated it now to the Electrum v4.0.9 server list. I mentioned the Electrum version in the commit message, do you think its sufficient?

LeoComandini replied:
Yes, sure

LeoComandini commented on subprojects/gdk_rust/gdk_electrum/src/lib.rs:
```diff
@@ -144,6 +152,12 @@ fn notify_fee(notif: NativeNotif, fees: &[FeeEstimate]) {
     notify(notif, data);
 }
+fn notify_updated_txs(notif: NativeNotif) {
+    // This is used as a signal to trigger syncina via get_transactions, the transaction
```
```suggestion
    // This is used as a signal to trigger syncing via get_transactions, the transaction
```

shesek replied:
Thanks, fixed in https://github.com/Blockstream/gdk/pull/59/commits/a903cb5d7f6723892621a5b668947ba8eaa1f6a9.

LeoComandini commented on subprojects/gdk_rust/tests/integration.rs:
```diff
@@ -119,3 +126,186 @@ fn liquid() {
+    assert_eq!(inv.longest_height, 145);
+    let txitem = test_session1.get_tx_from_list(&txid);
+    assert_eq!(txitem.block_height, 135);
+    assert_eq!(txitem.spv_verified, "verified");
```
```suggestion
    assert_eq!(txitem.spv_verified, "verified");

    test_session1.stop();
    test_session2.stop();    
```

LeoComandini commented:
> Rebased on master and removed the dependency on my `rust-electrum-client` fork (the necessary changes were added upstream and released in v0.4.0). It still depends on my `rust-bitcoin` fork pending a PR I sent there.
> 
> Other standing items/questions:
> 
>     * The apps need to handle the new `NotLongest` and `Unconfirmed` SPV verification statuses, as well as support configuring the new `spv_cross_validation` and `spv_cross_validation_servers` options.
> 
>     * Should the SPV cross-validation status be exposed as global session state, in addition to being included on individual transactions?
>       This could be useful for displaying cross-validation failures somewhere prominent (outside of individual txs), and could also provide the app some additional details about the chain fork, like the local height, the height of the best known chain and the additional work in the best chain.
> 
>     * Cross-validation is still not implemented in the C FFI used by the Green apps for SPV verification. In that environment, the Rust side doesn't run background threads of its own, it only runs when invoked from the CPP side. This introduces some complexities, both because cross-validation is an ongoing process and because the secondary servers are likely to be slow (or entirely unresponsive), which could result in user-visible latency if its not done in a separate thread.

Currently no production app is using SPV verification. The points you raise are reasonable and will need to be addressed.
But getting the interface changes using by Green apps right it's not easy or fast. We must make sure those work and have performances with all supported apps and platform. Such process will involve engineers, testers and designers and will take a rather long period. Therefore the plan is to postpone interface decisions and review and merge these SPV cross validation changes even though they might be unusable or hard to use from Green sessions.

LeoComandini commented:
DEBUG=1 ./launch_integration_tests.sh bitcoin
    DEBUG=1 ./launch_integration_tests.sh bitcoin

both fails with

    thread 'bitcoin' panicked at 'assertion failed: `(left == right)`
      left: `"unconfirmed"`,
     right: `"in_progress"`', tests/test_session.rs:451:9

I guess they need updating

LeoComandini commented:
Please add to `.gitlab-ci.yml`

    DEBUG=1 ./launch_integration_tests.sh spv_cross_validate
    DEBUG=1 ./launch_integration_tests.sh spv_cross_validation_session

LeoComandini commented on .gitlab-ci.yml:
```diff
@@ -180,7 +180,7 @@ test_rust:
     - cd $CI_PROJECT_DIR
     - git clone https://github.com/Blockstream/electrs electrs-elements
-    - cd electrs-elements && git checkout 5bae341585f70699cf12b587a1e9d392df43d674
+    - cd electrs-elements && git checkout d6043203310168bee18e47c19772ef375b9daaee
```
Should we upgrade to https://github.com/Blockstream/electrs/commit/1cbc86fbfdb5680fea752377bf8f18ed417f0edd?

shesek replied:
Updated to the latest (https://github.com/Blockstream/electrs/commit/65e20f9e1196c0039d56307b34ea7bcfbe375b99) in 70c9c23fe1934628672d05cb4fb41e71d085b8af.

shesek commented:
> I guess they need updating

Indeed, fixed in d903e0b04b027984494e1c04bcde76cb5f9aaedc.

shesek commented:
> Please add to `.gitlab-ci.yml`

Added in 31ebd845444b943b4bb843726b8ac2068afa7176.

LeoComandini commented on subprojects/gdk_rust/tests/integration.rs:
```diff
@@ -119,3 +126,189 @@ fn liquid() {
+// Test high-level session management, background validation and transaction status
+#[test]
+fn spv_cross_validation_session() {
+    let (mut test_session1, test_session2) = setup_forking_sessions(true);
```
```suggestion
    let (mut test_session1, mut test_session2) = setup_forking_sessions(true);
```
this is needed to stop the session, I forgot in the previous suggestion

shesek replied:
Fixed (squashed into the previous commit, now https://github.com/Blockstream/gdk/pull/59/commits/59b365494798491f3b7d103a2da7587ef219c84a)

LeoComandini commented on subprojects/gdk_rust/gdk_electrum/src/interface.rs:
```diff
@@ -41,27 +41,56 @@ pub struct WalletCtx {
-            ElectrumUrl::Plaintext(url) => {
-                let client = RawClient::new(&url)?;
-                Ok(Client::TCP(client))
+                (format!("tls://{}", url), config.validate_domain(*validate))
```
This must be `"ssl://{}"`

https://docs.rs/electrum-client/0.4.0-beta.1/electrum_client/client/struct.Client.html#method.new

shesek replied:
Good catch, thanks! Not the first time I get bitten by `tls` vs `ssl` 😅

Fixed in 6fdcd995d0df7b735721949996ec5cb9e184d2ee

LeoComandini commented:
ACK 6fdcd995d0df7b735721949996ec5cb9e184d2ee, code review, verified `server-*.txt`, gitlab CI passed, run tests locally, tested interface in testnet with python wheels (spv enabled cross validation disabled, cross validation enabled default secondary servers, cross validation enabled custom secondary servers, cross validation enabled secondary servers not working, cross validation enabled mixed secondary servers)

I guess we should wait until https://github.com/rust-bitcoin/rust-bitcoin/pull/511 is merged.

This should also be rebased, feel free to rebase multiple times, as for now most changes in master do not affect the Rust part so I can check easily the diff from the branch with the acked commit.

The points you raised above will probably be discussed and addressed once we'll be closer to have this feature exposed in the apps.

shesek commented:
@LeoComandini Excellent, thanks for the review!

I updated https://github.com/rust-bitcoin/rust-bitcoin/pull/511 just now to address some feedback, will send a ping here once that's gets merged/released.

shesek commented:
Rebased on master. Also added another commit, be7808ca858765d1fb1b4088486d4e4cf970218f, which is a refactor that shouldn't affect behavior.

LeoComandini commented on subprojects/gdk_rust/Cargo.toml:
```diff
@@ -30,12 +30,16 @@ ureq = { version = "1.0.0", features = ["json"] }
 bitcoincore-rpc = { git="https://github.com/romanz/rust-bitcoincore-rpc", rev="637e895a0b73ffc21c8f62fa2c65b6fd1247785a" }
-electrum-client = "0.3.0-beta.1"
-elements = { version = "0.13", features = ["serde-feature"] }
+electrum-client = "0.6.0-beta.1"
```
s/`0.6.0-beta.1`/`0.6.0`/ or maybe even `0.7.0`

shesek replied:
Updated (amended into 6d47005cd956101998ad38f478e8cf32d4fad603)

LeoComandini commented on subprojects/gdk_rust/gdk_common/src/network.rs:
```diff
@@ -26,6 +26,8 @@ pub struct Network {
     pub ct_exponent: Option<i32>,
     pub ct_min_value: Option<u64>,
     pub spv_enabled: Option<bool>,
+    pub spv_cross_validation: Option<bool>,
```
This unfortunately will cause the root dir for cache and store to change, thus the old store and its memos and settings will be ignored.

        let wallet_desc = format!("{}{:?}", xpub, self.network);
        let wallet_id = hex::encode(sha256::Hash::hash(wallet_desc.as_bytes()));
        ...
        let mut path: PathBuf = self.data_root.as_str().into();
        if !path.exists() {
            std::fs::create_dir_all(&path)?;
        }
        path.push(wallet_id);
        info!("Store root path: {:?}", path);

I think these new elements should be ignored when computing `wallet_desc`. Which is quite ugly, but if we want to limit the scope of this MR, we might need to accept a similar solution.

shesek replied:
Good catch! Addressed by c73f42779648e328b8cfa95a601ef085f43d6507.

LeoComandini commented on subprojects/gdk_rust/gdk_common/src/be/blockhash.rs:
```diff
@@ -0,0 +1,34 @@
+}
+
+// We must have a default for Store, so we use bitcoin::BlockHash which
+// will be replacted with the proper type after the first sync.
```
s/replacted/replaced/

shesek replied:
Fixed (amended into 6d47005cd956101998ad38f478e8cf32d4fad603)

LeoComandini commented:
nit: typo "rust-elemenets" in last commit

LeoComandini commented:
Also, is this still WIP?

LeoComandini commented:
re-ACK c73f42779648e328b8cfa95a601ef085f43d6507, review diff since last acked version, gitlab CI passed, manually tested that store data (settings and memos) are preserved after upgrade (`0.0.40` to c73f42779648e328b8cfa95a601ef085f43d6507) and downgrade.

LeoComandini commented:
Thanks @shesek!
